### PR TITLE
Fixed #24945 -- PostgreSQL fuzzystrmatch support.

### PR DESCRIPTION
**Pull request by dracos**
Just as I was in the area adding the trigram code, I thought I might as well take a look at this very similar functionality. I use Double Metaphone on my (Django powered) site http://theatricalia.com/ so that e.g. http://theatricalia.com/search?q=katie+stevens resolves to Katy Stephens.

### CODE REVIEW

**Comment by atombrella** [comment]
Could you add Levenhstein to docs/spelling_wordlist (you're probably also asked to rebase the commits into a single commit when you've done that) ? That should fix the failing test (docs).

**Comment by dracos** [comment]
@atombrella Thanks :)

**Comment by mjtamlyn** [comment]
Hmmm there's some decisions to be made here between the overlap of trigrams, fuzzy string match and full text search. Each have their own uses and they're probably all useful in their own way. If it's ok with you @dracos I'd like to not merge either this or #4815 at the moment - when I have #4726 a little more feature complete (it's getting there!) then I can start writing documentation which combines the collection of these features together, integrating in your code as well. It may be we want to lay things out slightly differently, with these functions appearing in `contrib.postgres.search` as well.

I'd also like to get some guidance from some "experts", probably via a post on django dev or similar as to the relative benefits and problems of these different approaches.

**Comment by dracos** [comment]
Yes, of course; I only really did this one as I had done #4815 and it seemed very similar (as you say) and was already in my head. I'll be sticking with an `extra()` for now anyway, got old versions to support :)

**Comment by timgraham** [comment]
I guess we'll close this for now until Marc's work is further along.

**Comment by imaia** [comment]
Could someone please link Marc's work here? How it is going? This PR is very important. Many people just have to install a apache lucene based solution because this is not available.

**Comment by charettes** [comment]
I also believe this should be re-opened as FTS search was merged without _fuzzy_ support. This will need to be rebased and the release note should be adjusted for 1.10.

**Comment by apollo13** [comment]
@imaia This PR is hardly "very" important -- if you want this feature, feel free to improve this PR to a state where it is mergeable and cleaned up.

**Comment by imaia** [comment]
@apollo13 because it is a feature? Although, nice suggestion. I'll probably be able to do it next week, if things go well for me. 
@charettes fair enough.

**Comment by imaia** [comment]
Any suggestion on how to handle test only models for fuzzystrmatch? Django 1.9 seems to not be handling the idea very well.

**Comment by timgraham** [comment]
If you're developing an updated patch, then please be more specific about what problem you're running into, even sending a work in progress PR would be okay. This won't be backported to 1.9. If you're asking some other usage question, then please use our support channels like the django-users mailing list.

**Comment by dracos** [comment]
This should be rebased and updated now.

**Comment by imaia** [comment]
Nice! @dracos what about adding LEVENSHTEIN_LESS_EQUAL support too, as it is a " accelerated version of levenshtein function for low values of distance"?
 
http://www.postgresql.org/docs/9.1/static/fuzzystrmatch.html

**Comment by dracos** [comment]
I've added it in a second commit, if that's considered okay.

File: tests/postgres_tests/test_fuzzystrmatch.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            [('Emelia', 1), ('Amelie', 1), ('Emily', 3), ('Fred', 4)],
+            transform=lambda instance: (instance.field, instance.distance),
+            ordered=True
+        )
+
+
+class FuzzyStrMatchTextFieldTest(FuzzyStrMatchTest):
+    """
+    TextField should have the exact same behavior as CharField
+    regarding trigram lookups.
```
Comment (mjtamlyn) [inline tests/postgres_tests/test_fuzzystrmatch.py]: Not trigrams here!

**Comment by mjtamlyn** [comment]
So, I'm less confident about the inclusion of these than I am with trigrams. I suppose what I'd really like to see is some discussion or examples of when these features are useful. Some questions which spring to mind:

- How do I choose between Trigram and Levenshtein?
- How do I choose between the 3 soundy matches?
- How do the sound matches work in not-english?

File: tests/postgres_tests/test_fuzzystrmatch.py (hunk 1 above)
Comment (dracos) [inline tests/postgres_tests/test_fuzzystrmatch.py]: Whoops, thanks!

**Comment by dracos** [comment]
Yeah, it's tricky, I don't know (I'm sure there must be papers).

The example I gave in the PR above is a Django project using double metaphone for when people search for names; I find it really useful. I think this is a good case because names are a good candidate for a 'sounds like' fuzzy matching algorithm.

I'd say you could only have double metaphone for a soundy algorithm and not bother with soundex (which is definitely rubbish) or metaphone, but OTOH metaphone at least is simpler with only the one output (I hadn't yet implemented the dmetaphone_alt function in this PR, but it is used on my site). Soundex and Metaphone are definitely English only; double metaphone takes account of other language origins with its alternate response, but still biased towards US pronunciation.

As for trigram/levenshtein, Levenshtein seems better suited to spot occasional typos/transpositions, whereas trigram seems more suited to a wider 'similarity' between two pieces of text, but I can't easily think of "in this scenario, use this one" type response. Trigram would work better if you had multiple words where their order could differ (but if you didn't care about the order, you could just sort them first…). You can't index levenshtein, as I understand it, which may make a difference; you could use levenshtein to help you weight/rank trigram filtered results, perhaps.

And of course there are other algorithms not in postgres, e.g. Beider-Morse phonetic matching.

**Comment by imaia** [comment]
I had some personal tryouts with Levenhstein, LevenhsteinLessEqual and Trigram. Within my tests, Trigram went very bad with short words. After some tunning, LevenhsteinLessEqual behaved the best in most scenarios.

**Comment by timgraham** [comment]
Where do we stand with this issue? Should we pose the remaining questions to the mailing list?

**Comment by mjtamlyn** [comment]
My instinct is that they are very simple to implement if you want to use them, and that it's hard to give good reasons as to why you should. I think a "django-postgres-functions" third party package could be a good home.

**Pull request closed without merging**